### ISSUE TITLE
Create ASM api and move asm off the server classpath

### ISSUE THREAD

**Issue by pgomulka** [elastic/elasticsearch#89806]
### Description

`org.ow2.asm:asm` dependency is used by `modules/lang-painless` `modules/lang-expression` and `x-pack/plugin/security`. It might be used by external plugins as well.
Introducing ASM dependency to server - needed for stable plugin scanning and registration - might cause problems for external "old style/unstable" plugins.

We should move ASM dependency into libs/asm and load it with EmbededImplClassLoader.

Should we refactor painless & expression to use the new SPI?

The altertnative is to upgrade asm version across the elasticsearch server, plugins and modules and rely on ASM bwc guarantees (it should be bwc compatible with older versions)

**Comment by ChrisHegarty** [comment]
We can follow a similar pattern that has been established with x-content/Jackson. That is, embedded loader impl can be used to load ASM for use by server (this requires a minimal API so that server does not use ASM directly). The painless and expression components can remain as are - they use ASM directly.

**Comment by tvernum** [comment]
I don't think it's reasonable to assume that all modules will align on a single ASM version. For example the dependency in security is there because the OpenID Connect library we use has a dependency on ASM. If we require everything in Elasticsearch to use the same version of ASM, then we are forced to run other dependencies with different ASM versions than the ones they officially support and are tested with.
It might work, but we'd be locking ourselves into guaranteeing it will always work, with no reasonable options if we ever hit an issue

> needed for stable plugin scanning and registration

Have we considered moving the whole of the plugin scanning into a dynamically loaded library instead? It sounds like that could be a smaller interface than trying to create a facade for ASM itself.

**Comment by pgomulka** [comment]
we discussed this on stable plugin sync on 6th sep.
we considered 3 options:
1 - keeping the dependency and aligning the version. This we quickly dismissed because of OpenID library used in security. We could align to the version used in security, but this would lock us into this version until that lib decides to upgrade
2 - shade asm library and use the shaded (package renamed) in server. This we also decided not to use as it will allow 3rd party plugins for dangerous code generation and manipulation
3 - hide asm related code into lib (libs/plugin-scanner) we concluded this is the best option. We will expose a limited api to allow to scan plugins for `@NamedComponent` and `@Extensible`. We will not try to expose ASM api, but rather a "plugin domain" api (methods for scanForPlugins etc)
